### PR TITLE
joh/thoughtless-haddock

### PR DESCRIPTION
**Pull request by jrieken**

### CODE REVIEW

File: src/vscode-dts/vscode.proposed.chatProvider.d.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+declare module 'vscode' {
+
+	export interface ChatResponseFragment {
+		index: number;
```
Comment (roblourens) [inline src/vscode-dts/vscode.proposed.chatProvider.d.ts:9]: Why is an index needed?

File: src/vscode-dts/vscode.proposed.chatProvider.d.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	/**
+	 * Represents a large language model that accepts ChatML messages and produces a streaming response
+	 */
+	export interface ChatResponseProvider {
+		provideChatResponse(messages: ChatMessage[], options: { [name: string]: any }, progress: Progress<ChatResponseFragment>, token: CancellationToken): Thenable<any>;
+	}
+
+	export interface ChatResponseProviderMetadata {
+		// TODO: add way to compute token count
+		name: string;
```
Comment (roblourens) [inline src/vscode-dts/vscode.proposed.chatProvider.d.ts:22]: Why isn't it just on the provider?

File: src/vs/workbench/contrib/chat/common/chatServiceImpl.ts [hunk 3]
```diff
@@ -473,7 +477,29 @@ export class ChatService extends Disposable implements IChatService {
...
+
+			if ((typeof resolvedCommand === 'string' && typeof message === 'string' && this.chatSlashCommandService.hasCommand(resolvedCommand))) {
+				// contributed slash commands
+				// TODO: spell this out in the UI
+				const history: IChatMessage[] = [];
+				for (const request of model.getRequests()) {
+					if (typeof request.message !== 'string' || !request.response) {
+						continue;
+					}
+					if (isMarkdownString(request.response.response)) {
```
Comment (roblourens) [inline src/vs/workbench/contrib/chat/common/chatServiceImpl.ts:491]: The type of this changed, this check will always fail

**Comment by roblourens** [comment]
Fine with me to merge whenever it's ready ✅

File: src/vscode-dts/vscode.proposed.chatSlashCommands.d.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	export interface SlashResponse {
+		message: MarkdownString;
+		// edits?: TextEdit[] | WorkspaceEdit;
+	}
+
+	export interface SlashResult {
+		// followUp?: InteractiveSessionFollowup[];
+	}
+
+	export interface SlashCommand {
```
Comment (roblourens) [inline src/vscode-dts/vscode.proposed.chatSlashCommands.d.ts:35]: Will at least need a description for the slash command. There are also some other metadata fields that Joyce added https://github.com/microsoft/vscode/blob/af667558efb055a047080722900c500107a4f7c5/src/vscode-dts/vscode.proposed.interactive.d.ts#L155-L158

Comment (roblourens) [reply src/vscode-dts/vscode.proposed.chatSlashCommands.d.ts:35]: Actually I guess I'm not sure if this is expected to replace the old `provideSlashCommands` on the chat provider or if you are _only_ thinking about contributions from other extensions here

File: src/vs/workbench/contrib/chat/common/chatServiceImpl.ts (hunk 3 above)
Comment (jrieken) [reply src/vs/workbench/contrib/chat/common/chatServiceImpl.ts:491]: thanks

File: src/vscode-dts/vscode.proposed.chatProvider.d.ts (hunk 1 above)
Comment (jrieken) [reply src/vscode-dts/vscode.proposed.chatProvider.d.ts:9]: It's the choice things. So, you have two choices and fragments like `(1, Hello), (1, World), (0, Salve), (1, !)` for  "Salve" and "Hello World!"

This needs some proper API, like a ResponseItem which emits events for its chunks

File: src/vscode-dts/vscode.proposed.chatSlashCommands.d.ts (hunk 4 above)
Comment (jrieken) [reply src/vscode-dts/vscode.proposed.chatSlashCommands.d.ts:35]: For now it's just extension contributed commands but we can extend this as we go. The description/detail property exists but as of today it is defined statically inside package.json. This needs to change for dynamic commands, is actually already implemented like this but I forgot expose it here

**Review by bpasero — DISMISSED** [review]

**Review by bpasero — APPROVED** [review]

**Pull request merged**